File: backend/session_manager.py
import uuid
from models import SessionState


_store: dict[str, SessionState] = {}


def create_session() -> SessionState:
    session_id = str(uuid.uuid4())
    session = SessionState(session_id=session_id)
    _store[session_id] = session
    return session


def get_session(session_id: str):
    return _store.get(session_id)


def update_session(session_id: str, **kwargs):
    session = _store.get(session_id)
    if not session:
        return None
    for key, value in kwargs.items():
        if hasattr(session, key):
            setattr(session, key, value)
    return session


def end_session(session_id: str) -> None:
    session = _store.get(session_id)
    if session:
        session.status = "ended"

File: apps/api/src/session-manager.ts
import { randomUUID } from "node:crypto";
import {
  createInitialSessionState,
  type SessionState,
} from "@emergency-guidance/shared";

const store = new Map<string, SessionState>();

export function createSession(): SessionState {
  const id = randomUUID();
  const state = createInitialSessionState(id);
  store.set(id, state);
  return state;
}

export function getSession(sessionId: string): SessionState | undefined {
  return store.get(sessionId);
}

export function updateSession(
  sessionId: string,
  patch: Partial<SessionState>
): SessionState | undefined {
  const prev = store.get(sessionId);
  if (!prev) return undefined;
  const next = { ...prev, ...patch };
  store.set(sessionId, next);
  return next;
}

export function endSession(sessionId: string): void {
  const s = store.get(sessionId);
  if (s) store.set(sessionId, { ...s, status: "ended" });
}
